haystack/components/evaluators/document_recall.py

```diff
@@ -0,0 +1,112 @@
+import itertools
+from enum import Enum
+from typing import Any, Dict, List, Union
+
+from haystack.core.component import component
+from haystack.dataclasses import Document
+
+
+class RecallMode(Enum):
+    """
+    Enum for the mode to use for calculating the recall score.
+    """
+
+    SINGLE_HIT = "single_hit"
+    MULTI_HIT = "multi_hit"
+
+    def __str__(self):
+        return self.value
+
+    @staticmethod
+    def from_str(string: str):
+        enum_map = {e.value: e for e in RecallMode}
+        mode = enum_map.get(string)
+        if mode is None:
+            msg = f"Unknown mode '{string}'. Supported modes are: {list(enum_map.keys())}"
+            raise ValueError(msg)
+        return mode
+
+
+@component
+class DocumentRecallEvaluator:
+    """
+    Evaluator that calculates the Recall score for a list of questions.
+    Returns both a list of scores for each question and the average.
+    Each question can have multiple ground truth documents and multiple predicted documents.
+
+    Usage example:
+    ```python
+    from haystack.components.evaluators import DocumentRecallEvaluator
+    evaluator = DocumentRecallEvaluator()
+    result = evaluator.run(
+        questions=["What is the capital of Germany?", "What is the capital of France?"],
+        ground_truth_answers=[["Berlin"], ["Paris"]],
+        predicted_answers=[["Paris"], ["London"]],
+    )
+    print(result["individual_scores"])
+    # [0.0, 0.0]
+    print(result["score"])
+    # 0.0
+    ```
+    """
+
+    def __init__(self, mode: Union[str, RecallMode] = RecallMode.SINGLE_HIT):
+        """
+        Create a DocumentRecallEvaluator component.
+
+        :param mode:
+            Mode to use for calculating the recall score.
+        """
+        if isinstance(mode, str):
+            mode = RecallMode.from_str(mode)
+
+        mode_functions = {RecallMode.SINGLE_HIT: self._recall_single_hit, RecallMode.MULTI_HIT: self._recall_multi_hit}
+        self.mode_function = mode_functions[mode]
+
+    def _recall_single_hit(self, ground_truth_documents: List[Document], retrieved_documents: List[Document]) -> float:
+        retrieved_ground_truths = {
+            g.id
+            for g, p in itertools.product(ground_truth_documents, retrieved_documents)
+            if g.content and p.content and g.content == p.content
+        }
+        return len(retrieved_ground_truths) / len(ground_truth_documents)
```

Comment:
This seems like it does the same thing as the multi hit function below? For single hit, you just need to return true if any of the GT docs appear in the retrieved docs.